### DISCUSSION TITLE
API creating transaction results in error 422

### DISCUSSION THREAD

**Discussion by landon-lengyel** [firefly-iii/firefly-iii#9440] [Firefly III]
### Support guidelines

- [x] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [x] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [x] ... there are no open or closed issues that are related to my problem
- [ ] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

When I attempt to create a transaction with the API, I am receiving an odd error. I cannot figure this out. It's likely something I'm doing incorrectly, but I believe I'm following the documentation properly.

Error:
`{"jsonapi":{"version":"1.0"},"errors":[{"detail":"Need at least one transaction.","source":{"pointer":"\/transactions\/0\/description"},"status":"422"}]}`

With debug enabled, I see:
```
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Now in FireflyIII\Support\Request\ChecksLogin::authorize  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Request class has no acceptedRoles array  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Collect rules of TransactionStoreRequest  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Now in getTransactionsArray  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Now going to validateOneTransaction  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Now in validateOneTransaction  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Now in getTransactionsArray  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Added error: at_least_one_transaction.  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Now done with validateOneTransaction  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Now in getTransactionsArray  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Now going to validateOneTransaction  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Now in validateOneTransaction  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Now in getTransactionsArray  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Added error: at_least_one_transaction.  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Now done with validateOneTransaction  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Now in Handler::render()  
firefly_iii_core  | [2024-11-05 18:26:34] production.DEBUG: Return to parent to handle LaravelValidationException(422)  
firefly_iii_core  | 10.0.35.17 - - [05/Nov/2024:18:26:34 -0700] "POST /api/v1/transactions HTTP/1.1" 422 1064 "-" "Mozilla/5.0 (Linux; Freedesktop SDK 24.08 (Flatpak runtime); en-US) PowerShell/7.4.6"
```

### Debug information

Debug information generated at 2024-11-05 16:35:13 for Firefly III version **v6.1.21**.

<table><thead><tr><th colspan="2">System information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Firefly III</td><td>6.1.21 / v2.1.0 / 24 (exp. 24)</td></tr><tr><td>PHP version</td><td>8.3.12 (64bits) / apache2handler / Linux x86_64</td></tr><tr><td>BCscale</td><td>12</td></tr><tr><td>Error reporting</td><td>Display: Off, reporting: ALL errors</td></tr><tr><td>Max upload</td><td>67108864 (64 MB)</td></tr><tr><td>Database drivers</td><td><strong>*mysql*</strong>, pgsql, sqlite, </td></tr><tr><td>Docker build</td><td><span>#</span>1131, base <span>#</span>91</td></tr></tbody></table><table><thead><tr><th colspan="2">Firefly III information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Timezone</td><td>America/Denver + America/Denver</td></tr><tr><td>App environment</td><td>production, debug: true</td></tr><tr><td>Layout</td><td>v1</td></tr><tr><td>Logging</td><td>debug, stack / (empty)</td></tr><tr><td>Cache driver</td><td>file</td></tr><tr><td>Default language and locale</td><td>en_US + equal</td></tr><tr><td>Trusted proxies</td><td></td></tr><tr><td>Login provider & user guard</td><td>eloquent / web</td></tr><tr><td>Login headers</td><td>N/A + N/A</td></tr><tr><td>Stateful domains</td><td></td></tr><tr><td>Last cron job</td><td>never (never)</td></tr><tr><td>Mailer</td><td>log</td></tr></tbody></table><table><thead><tr><th colspan="2">User-specific information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>User</td><td><span>#</span>1 of 1</td></tr><tr><td>User flags</td><td><span title="Has piggy banks">:pig:</span></td></tr><tr><td>Session start</td><td>2024-11-01 00:00:00</td></tr><tr><td>Session end</td><td>2024-11-30 23:59:59</td></tr><tr><td>View range</td><td>1M</td></tr><tr><td>User language</td><td>en_US</td></tr><tr><td>User locale</td><td>en_US</td></tr><tr><td>Locale(s) supported</td><td>en_US.utf8: :white_check_mark:<br>en_US.UTF-8: :white_check_mark:<br></td></tr><tr><td>User agent</td><td>Mozilla/5.0 (X11; Linux x86_64; rv:132.0) Gecko/20100101 Firefox/132.0</td></tr></tbody></table>

### Expected behaviour

Transaction is created

### Steps to reproduce

Send POST request with the following parameters:

API Endpoint: `{{URL}}/api/v1/transactions`
Body: 
```json
{
  "apply_rules": false,
  "transactions": [
    {
      "date": "2024-04-07T00:00:00",
      "description": "BackBlaze",
      "type": "withdrawal",
      "amount": "3.21",
      "destination_name": "Electronics",
      "source_name": "Checking Account"
    }
  ]
}
```
Header:
```json
{
  "Authorization": "Bearer {{Key}}",
  "Accept": "application/vnd.api+json"
}
```

**Comment by JC5** [comment]
Sorry, I can't replicate this. Submitting the body from your issue works as expected.

```
curl --location 'https://firefly.sd.internal/api/v1/transactions' \
--header 'Accept: application/json' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer eyJ0eX...' \
--data '{
  "apply_rules": false,
  "transactions": [
    {
      "date": "2024-04-07T00:00:00",
      "description": "BackBlaze",
      "type": "withdrawal",
      "amount": "3.21",
      "destination_name": "Electronics",
      "source_name": "Checking Account"
    }
  ]
}'
```

**Comment by landon-lengyel** [comment]
Thank you for your quick response and trying it.

It looks like curl also works fine for me. I was trying to use PowerShell. I'm not sure what it could be doing to mangle it. Not an issue with Firefly by the looks of it.

If you leave this open I'll try to figure it out in case anyone else is using PowerShell, then close it.

**Comment by landon-lengyel** [comment]
Alright I did a packet capture and was able to find the issue. Both Invoke-RestMethod and Invoke-WebRequest do not include the header `Content-Type:application/json` by default, and the 422 error is caused by that. 

As far as Firefly goes, it would be nice if the error was more accurately attributed to the headers, and didn't refer to the description. 

Thanks again for looking at it with me!

**Comment by JC5** [comment]
But it does, I'm not sure you identified the problem correctly.

```
curl --location 'https://firefly.sd.internal/api/v1/transactions' \
--header 'Authorization: Bearer eyJ...' \
--data '{
  "apply_rules": false,
  "transactions": [
    {
      "date": "2024-04-07T00:00:00+04:00",
      "description": "BackBlaze",
      "type": "withdrawal",
      "amount": "3.21",
      "destination_name": "Electronics",
      "source_name": "Checking Account"
    }
  ]
}'
```

Result:

```
415 Unsupported Media Type
{
    "message": "Content-Type header cannot be empty.",
    "exception": "BadHttpHeaderException"
}
```

**Comment by MarcelRobitaille** [comment]
I got the same error message with python requests library. I set the content type to json, but I used the `data=` keyword argument so the body was actually url encoded. I fixed it by using the `json=` keyword argument. My fault, but a horrible error message for "your body is not actually json, idiot".
